// @inovua/reactdatagrid-community/packages/TextInput/examples/editors/index.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _react = _interopRequireWildcard(require("react"));

var _src = _interopRequireDefault(require("../../src"));

require("../../style/index.scss");

var _TextArea = _interopRequireDefault(require("../../../TextArea"));

require("../../../TextArea/style/index.scss");

var _NumericInput = _interopRequireDefault(require("../../../NumericInput"));

require("../../../NumericInput/style/index.scss");

var _DateInput = _interopRequireDefault(require("../../../Calendar/DateInput"));

require("../../../Calendar/style/index.scss");

var _ComboBox = _interopRequireDefault(require("../../../ComboBox"));

require("../../../ComboBox/style/index.scss");

var _countries = _interopRequireDefault(require("../../../ComboBox/examples/default/countries"));

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var EditorsExample = function (_Component) {
  (0, _inherits2.default)(EditorsExample, _Component);

  var _super = _createSuper(EditorsExample);

  function EditorsExample(props) {
    var _this;

    (0, _classCallCheck2.default)(this, EditorsExample);
    _this = _super.call(this, props);
    _this.state = {
      textInputValue: 'Text',
      textAreaValue: 'Content',
      numericInputValue: '2345',
      dateInputValue: '12-12-2017',
      comboValue: 'Albania',
      acceptClearToolFocus: false
    };
    return _this;
  }

  (0, _createClass2.default)(EditorsExample, [{
    key: "render",
    value: function render() {
      var _this2 = this;

      return _react.default.createElement("div", {
        style: {
          margin: 30
        }
      }, _react.default.createElement("div", {
        style: {
          marginBottom: 20
        }
      }, _react.default.createElement("input", {
        type: "checkbox",
        checked: this.state.acceptClearToolFocus,
        onChange: function onChange(ev) {
          return _this2.setState({
            acceptClearToolFocus: ev.target.checked
          });
        }
      }), "acceptClearToolFocus"), _react.default.createElement("table", null, _react.default.createElement("tbody", null, _react.default.createElement("tr", null, _react.default.createElement("td", null, "TextInput:"), _react.default.createElement("td", null, _react.default.createElement(_src.default, {
        style: {
          width: 230
        },
        value: this.state.textInputValue,
        onChange: function onChange(textInputValue) {
          return _this2.setState({
            textInputValue: textInputValue
          });
        },
        acceptClearToolFocus: this.state.acceptClearToolFocus
      }))), _react.default.createElement("tr", null, _react.default.createElement("td", null, "TextArea:"), _react.default.createElement("td", null, _react.default.createElement(_TextArea.default, {
        style: {
          height: 200,
          width: 230
        },
        value: this.state.textAreaValue,
        onChange: function onChange(textAreaValue) {
          return _this2.setState({
            textAreaValue: textAreaValue
          });
        }
      }))), _react.default.createElement("tr", null, _react.default.createElement("td", null, "NumericInput:"), _react.default.createElement("td", null, _react.default.createElement(_NumericInput.default, {
        style: {
          width: 230
        },
        value: this.state.numericInputValue,
        onChange: function onChange(numericInputValue) {
          return _this2.setState({
            numericInputValue: numericInputValue
          });
        },
        acceptClearToolFocus: this.state.acceptClearToolFocus
      }))), _react.default.createElement("tr", null, _react.default.createElement("td", null, "DateInput:"), _react.default.createElement("td", null, _react.default.createElement(_DateInput.default, {
        dateFormat: "DD-MM-YYYY hh:mm:ss",
        style: {
          width: 230
        },
        text: this.state.dateInputValue,
        onTextChange: function onTextChange(dateInputValue) {
          return _this2.setState({
            dateInputValue: dateInputValue
          });
        },
        showClock: true
      }))), _react.default.createElement("tr", null, _react.default.createElement("td", null, "ComboBox:"), _react.default.createElement("td", null, _react.default.createElement(_ComboBox.default, {
        inlineFlex: true,
        dataSource: _countries.default,
        style: {
          width: 230
        },
        value: this.state.comboValue,
        onChange: function onChange(comboValue) {
          return _this2.setState({
            comboValue: comboValue
          });
        }
      }))))));
    }
  }]);
  return EditorsExample;
}(_react.Component);

var _default = EditorsExample;
exports.default = _default;
