// @inovua/reactdatagrid-community/common/testUtils.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.render = exports.default = void 0;
exports.simulateKeyboardEvent = simulateKeyboardEvent;
exports.simulateMouseEvent = simulateMouseEvent;

var _reactDom = _interopRequireDefault(require("react-dom"));

var _getGlobal = require("../getGlobal");

var globalObject = (0, _getGlobal.getGlobal)();

var render = function render(cmp) {
  var targetNode = document.createElement('div');
  targetNode.style.height = '600px';
  targetNode.style.width = '1000px';
  document.body.appendChild(targetNode);

  var instance = _reactDom.default.render(cmp, targetNode);

  instance.rerender = function (cmp) {
    return _reactDom.default.render(cmp, targetNode);
  };

  instance.unmount = function () {
    _reactDom.default.unmountComponentAtNode(targetNode);

    document.body.removeChild(targetNode);
  };

  return instance;
};

exports.render = render;
var _default = render;
exports.default = _default;

function simulateMouseEvent(eventType, target) {
  var evt = new MouseEvent(eventType, {
    bubbles: true,
    cancelable: true,
    view: globalObject
  });
  target.dispatchEvent(evt);
}

function simulateKeyboardEvent(eventType, target, key) {
  var evt = new KeyboardEvent(eventType, {
    bubbles: true,
    cancelable: true,
    view: window,
    key: key
  });
  target.dispatchEvent(evt);
}
