// @inovua/reactdatagrid-community/NumberFilter/index.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _NumericInput = _interopRequireDefault(require("../packages/NumericInput"));

var _lodash = _interopRequireDefault(require("lodash.debounce"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var NumberFilter = function (_React$Component) {
  (0, _inherits2.default)(NumberFilter, _React$Component);

  var _super = _createSuper(NumberFilter);

  function NumberFilter(props) {
    var _this;

    (0, _classCallCheck2.default)(this, NumberFilter);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "input", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "refInput", void 0);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "componentDidUpdate", function (_ref) {
      var value = _ref.filterValue.value;

      if (_this.props.filterValue) {
        if (value < _this.props.filterValue && _this.props.filterValue.value || value > _this.props.filterValue && _this.props.filterValue.value) {
          _this.setValue(_this.props.filterValue && _this.props.filterValue.value);
        }
      }
    });

    _this.refInput = function (i) {
      _this.input = i;
    };

    _this.state = {
      value: props.filterValue ? props.filterValue.value || '' : ''
    };
    _this.onChange = _this.onChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.onStartChange = _this.onStartChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.onEndChange = _this.onEndChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.onValueChange = _this.onValueChange.bind((0, _assertThisInitialized2.default)(_this));
    var filterDelay = props.filterDelay;

    if (filterDelay && filterDelay >= 1) {
      _this.onValueChange = (0, _lodash.default)(_this.onValueChange, filterDelay, {
        leading: false,
        trailing: true
      });
    }

    return _this;
  }

  (0, _createClass2.default)(NumberFilter, [{
    key: "onChange",
    value: function onChange(value) {
      if (value === this.state.value) {
        return;
      }

      this.onValueChange(value);
      this.setValue(value);
    }
  }, {
    key: "onStartChange",
    value: function onStartChange(start) {
      if (this.state.value) {
        if (this.state.value.start && start === this.state.value.start) {
          return;
        }
      }

      var newValue = _objectSpread(_objectSpread({}, this.state.value), {}, {
        start: start
      });

      this.onValueChange(newValue);
      this.setValue(newValue);
    }
  }, {
    key: "onEndChange",
    value: function onEndChange(end) {
      if (this.state.value) {
        if (this.state.value.end && end === this.state.value.end) {
          return;
        }
      }

      var newValue = _objectSpread(_objectSpread({}, this.state.value), {}, {
        end: end
      });

      this.onValueChange(newValue);
      this.setValue(newValue);
    }
  }, {
    key: "setValue",
    value: function setValue(value) {
      this.setState({
        value: value
      });
    }
  }, {
    key: "onValueChange",
    value: function onValueChange(value) {
      this.props.onChange && this.props.onChange(_objectSpread(_objectSpread({}, this.props.filterValue), {}, {
        value: value
      }));
    }
  }, {
    key: "render",
    value: function render() {
      var _this$props = this.props,
          filterValue = _this$props.filterValue,
          i18n = _this$props.i18n,
          filterEditorProps = _this$props.filterEditorProps;
      var _this$props2 = this.props,
          readOnly = _this$props2.readOnly,
          disabled = _this$props2.disabled,
          theme = _this$props2.theme;

      if (filterEditorProps == null) {
        filterEditorProps = filterValue && filterValue.filterEditorProps;
      }

      var inputProps = {
        readOnly: readOnly,
        disabled: disabled,
        theme: theme,
        style: {
          minWidth: 0
        }
      };

      if (filterValue) {
        inputProps.value = this.state.value;
      }

      switch (filterValue && filterValue.operator) {
        case 'inrange':
        case 'notinrange':
          var _ref2 = this.state.value || {
            start: '',
            end: ''
          },
              start = _ref2.start,
              end = _ref2.end,
              startInputProps = _objectSpread(_objectSpread({}, inputProps), {}, {
            value: start
          }),
              endInputProps = _objectSpread(_objectSpread({}, inputProps), {}, {
            value: end
          });

          var startFilterEditorProps = typeof filterEditorProps === 'function' ? filterEditorProps(this.props, {
            value: start,
            index: 0
          }) : filterEditorProps;
          var endFilterEditorProps = typeof filterEditorProps === 'function' ? filterEditorProps(this.props, {
            value: end,
            index: 1
          }) : filterEditorProps;

          var startProps = _objectSpread(_objectSpread({
            placeholder: i18n && i18n('start')
          }, startFilterEditorProps), {}, {
            ref: this.refInput,
            onChange: this.onStartChange,
            className: 'InovuaReactDataGrid__column-header__filter InovuaReactDataGrid__column-header__filter--number'
          }, startInputProps);

          var endProps = _objectSpread(_objectSpread({
            placeholder: i18n && i18n('end')
          }, endFilterEditorProps), {}, {
            ref: this.refInput,
            onChange: this.onEndChange,
            className: 'InovuaReactDataGrid__column-header__filter InovuaReactDataGrid__column-header__filter--number'
          }, endInputProps);

          return this.props.render && this.props.render(_react.default.createElement("div", {
            style: {
              display: 'flex'
            }
          }, _react.default.createElement(_NumericInput.default, _objectSpread({}, startProps)), _react.default.createElement("div", {
            className: "InovuaReactDataGrid__column-header__filter__binary_operator_separator"
          }), _react.default.createElement(_NumericInput.default, _objectSpread({}, endProps))));

        default:
          var finalEditorProps = typeof filterEditorProps === 'function' ? filterEditorProps(this.props, {
            value: this.state.value
          }) : filterEditorProps;

          var finalProps = _objectSpread(_objectSpread({}, finalEditorProps), {}, {
            ref: this.refInput,
            onChange: this.onChange,
            arrowSize: 10,
            className: 'InovuaReactDataGrid__column-header__filter InovuaReactDataGrid__column-header__filter--number'
          }, inputProps);

          return this.props.render && this.props.render(_react.default.createElement(_NumericInput.default, _objectSpread({}, finalProps)));
      }
    }
  }]);
  return NumberFilter;
}(_react.default.Component);

var _default = NumberFilter;
exports.default = _default;
