// @inovua/reactdatagrid-community/packages/ToolBar/examples/ResizableContainer.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _Panel = _interopRequireDefault(require("../../Panel"));

require("../../Panel/style/index.scss");

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var _default = function _default(props) {
  var renderBody = function renderBody() {
    return _react.default.createElement("div", {
      style: {
        padding: '0 20px'
      }
    }, _react.default.createElement("p", null, "When we started building the toolkit, we've made a checklist of features that our components need to include out-of-the-box:"), _react.default.createElement("ul", null, _react.default.createElement("li", null, _react.default.createElement("b", null, "Performance"), " - a component is only useful if it does its job quickly. This will generally not be a problem with smaller components like buttons, dialogs, color pickers, etc - but menus, lists and grids need a lot of performance considerations in order to be really snappy.")));
  };

  var width = props.width !== undefined ? props.width : '100%';
  return _react.default.createElement(_Panel.default, _objectSpread(_objectSpread({}, props), {}, {
    style: {
      maxWidth: width
    },
    bodyStyle: {
      paddingTop: 30
    },
    renderTitleBar: function renderTitleBar() {
      return props.children;
    },
    renderBody: renderBody,
    bodyScrollable: true
  }));
};

exports.default = _default;
