// @inovua/reactdatagrid-community/testUtils.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
Object.defineProperty(exports, "findBodyCells", {
  enumerable: true,
  get: function get() {
    return _utils.findBodyCells;
  }
});
Object.defineProperty(exports, "findCells", {
  enumerable: true,
  get: function get() {
    return _utils.findCells;
  }
});
Object.defineProperty(exports, "findHeaderCells", {
  enumerable: true,
  get: function get() {
    return _utils.findHeaderCells;
  }
});
Object.defineProperty(exports, "findRowAt", {
  enumerable: true,
  get: function get() {
    return _utils.findRowAt;
  }
});
Object.defineProperty(exports, "findRows", {
  enumerable: true,
  get: function get() {
    return _utils.findRows;
  }
});
Object.defineProperty(exports, "getSortIcons", {
  enumerable: true,
  get: function get() {
    return _utils.getSortIcons;
  }
});
exports.render = void 0;
exports.simulateKeyboardEvent = simulateKeyboardEvent;
exports.simulateMouseEvent = simulateMouseEvent;
Object.defineProperty(exports, "wait", {
  enumerable: true,
  get: function get() {
    return _utils.wait;
  }
});

var _reactDom = _interopRequireDefault(require("react-dom"));

var _react = require("react");

var _utils = require("./__test__/utils");

var _getGlobal = require("./getGlobal");

var globalObject = (0, _getGlobal.getGlobal)();

var render = function render(cmp) {
  var targetNode = document.createElement('div');
  targetNode.style.height = '600px';
  targetNode.style.width = '1000px';
  document.body.appendChild(targetNode);

  var instance = _reactDom.default.render(cmp, targetNode);

  instance.rerender = function (cmp) {
    return _reactDom.default.render(cmp, targetNode);
  };

  instance.setProps = function (newProps) {
    return _reactDom.default.render((0, _react.cloneElement)(cmp, newProps), targetNode);
  };

  instance.unmount = function () {
    _reactDom.default.unmountComponentAtNode(targetNode);

    document.body.removeChild(targetNode);
  };

  return instance;
};

exports.render = render;
var _default = render;
exports.default = _default;

function simulateMouseEvent(eventType, target) {
  var evt = new MouseEvent(eventType, {
    bubbles: true,
    cancelable: true,
    view: globalObject
  });
  target.dispatchEvent(evt);
}

function simulateKeyboardEvent(eventType, target, key) {
  var evt = new KeyboardEvent(eventType, {
    bubbles: true,
    cancelable: true,
    view: window,
    key: key
  });
  target.dispatchEvent(evt);
}
