// @inovua/reactdatagrid-community/getColumnState/defaultRowExpandColumn.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
Object.defineProperty(exports, "rowExpandColumnId", {
  enumerable: true,
  get: function get() {
    return _defaultRowExpandColumnId.id;
  }
});

var _react = _interopRequireDefault(require("react"));

var _defaultRowExpandColumnId = require("./defaultRowExpandColumnId");

var ICON_EXPANDED = _react.default.createElement("svg", {
  width: "24",
  height: "24",
  viewBox: "0 0 24 24"
}, _react.default.createElement("path", {
  d: "M19 13H5v-2h14v2z"
}));

var ICON_COLLAPSED = _react.default.createElement("svg", {
  width: "24",
  height: "24",
  viewBox: "0 0 24 24"
}, _react.default.createElement("path", {
  d: "M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
}));

var ICON_MORE = _react.default.createElement("svg", {
  width: "24",
  height: "24",
  viewBox: "0 0 24 24",
  style: {
    position: 'relative',
    top: 3
  }
}, _react.default.createElement("path", {
  d: "M6 10c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm12 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2zm-6 0c-1.1 0-2 .9-2 2s.9 2 2 2 2-.9 2-2-.9-2-2-2z"
}));

var _default = {
  id: _defaultRowExpandColumnId.id,
  rowExpandColumn: true,
  cellSelectable: false,
  headerAlign: 'center',
  textAlign: 'center',
  render: function render(_ref) {
    var isRowExpandable = _ref.isRowExpandable,
        rowExpanded = _ref.rowExpanded,
        toggleRowExpand = _ref.toggleRowExpand;

    if (!isRowExpandable()) {
      return;
    }

    var style = {
      cursor: 'pointer'
    };
    return _react.default.cloneElement(rowExpanded ? ICON_EXPANDED : ICON_COLLAPSED, {
      style: style,
      onClick: function onClick(event) {
        event.stopPropagation();
        toggleRowExpand();
      }
    });
  },
  header: ICON_MORE,
  showInContextMenu: false,
  showColumnMenuSortOptions: false,
  showColumnMenuGroupOptions: false,
  showColumnMenuTool: false,
  sortable: false,
  editable: false,
  groupBy: false,
  defaultWidth: 50,
  minWidth: 40
};
exports.default = _default;
