// @inovua/reactdatagrid-community/packages/Overlay/src/getPosition/getPosition.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _regionAlign = _interopRequireDefault(require("../../../packages/region-align"));

var _assign = _interopRequireDefault(require("../../../common/assign"));

var _getPositionOffsets = _interopRequireDefault(require("./getPositionOffsets"));

var _positionsMap = _interopRequireDefault(require("../positionsMap"));

var _getArrowPosition = _interopRequireDefault(require("./getArrowPosition"));

var _getConstrainRegion = _interopRequireDefault(require("../../../common/getConstrainRegion"));

function getPosition(_ref) {
  var _ref$constrainTo = _ref.constrainTo,
      constrainTo = _ref$constrainTo === void 0 ? true : _ref$constrainTo,
      targetNode = _ref.targetNode,
      overlayNode = _ref.overlayNode,
      offset = _ref.offset,
      positions = _ref.positions,
      relativeToViewport = _ref.relativeToViewport,
      arrowSize = _ref.arrowSize;

  if (!constrainTo || !overlayNode || !targetNode) {
    return null;
  }

  var constrain = (0, _getConstrainRegion.default)(constrainTo, targetNode);

  var overlayRegion = _regionAlign.default.from(overlayNode);

  var alignRegion = _regionAlign.default.from(targetNode);

  var newRegion = overlayRegion.clone();
  var positionsNormalized = positions.map(function (position) {
    return _positionsMap.default[position].position;
  });
  var positionsOffsets = (0, _getPositionOffsets.default)(positions, offset);
  var positionsOffsetsClone = positionsOffsets.map(function (offset) {
    return (0, _assign.default)({}, offset);
  });
  var succesfullPosition = newRegion.alignTo(alignRegion, positionsNormalized, {
    constrain: constrain,
    offset: positionsOffsetsClone
  });
  var arrowConfig = (0, _getArrowPosition.default)({
    arrowSize: arrowSize,
    overlayRegion: newRegion,
    targetRegion: alignRegion,
    position: succesfullPosition
  });
  var position = {
    top: newRegion.top,
    left: newRegion.left
  };

  if (!relativeToViewport && overlayNode.offsetParent) {
    var offsetParentRegion = _regionAlign.default.from(overlayNode.offsetParent);

    position.left -= offsetParentRegion.left;
    position.top -= offsetParentRegion.top;
  }

  return {
    arrowConfig: arrowConfig,
    position: position,
    alignRegion: alignRegion
  };
}

var _default = getPosition;
exports.default = _default;
