// @inovua/reactdatagrid-community/packages/ComboBox/examples/default/index.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

var _typeof = require("@babel/runtime/helpers/typeof");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireWildcard(require("react"));

require("../../style/index.scss");

require("./index.scss");

var _countries = _interopRequireDefault(require("./countries"));

var _ComboBox = _interopRequireDefault(require("../../src/ComboBox"));

require("../../../CheckBox/style/index.scss");

require("../../../NumericInput/style/index.scss");

require("../../../TextInput/style/index.scss");

var _timers = require("timers");

function _getRequireWildcardCache(nodeInterop) { if (typeof WeakMap !== "function") return null; var cacheBabelInterop = new WeakMap(); var cacheNodeInterop = new WeakMap(); return (_getRequireWildcardCache = function _getRequireWildcardCache(nodeInterop) { return nodeInterop ? cacheNodeInterop : cacheBabelInterop; })(nodeInterop); }

function _interopRequireWildcard(obj, nodeInterop) { if (!nodeInterop && obj && obj.__esModule) { return obj; } if (obj === null || _typeof(obj) !== "object" && typeof obj !== "function") { return { default: obj }; } var cache = _getRequireWildcardCache(nodeInterop); if (cache && cache.has(obj)) { return cache.get(obj); } var newObj = {}; var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor; for (var key in obj) { if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) { var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null; if (desc && (desc.get || desc.set)) { Object.defineProperty(newObj, key, desc); } else { newObj[key] = obj[key]; } } } newObj.default = obj; if (cache) { cache.set(obj, newObj); } return newObj; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

var themes = ['default', 'amber-800', 'green-800', 'red-800'].map(function (name, index) {
  return {
    label: name,
    id: index
  };
});
var testDataSource = [{
  id: 'test',
  label: 'a very long text comes in here that will cause'
}, {
  id: 'test2',
  label: 'text2'
}, {
  id: 'test3',
  label: 'text3'
}, {
  id: 7,
  label: 'seven'
}, {
  id: true,
  label: 'true'
}, {
  id: false,
  label: 'false'
}];

var namedCountries = _countries.default.map(function (c) {
  return {
    id: c.id,
    name: c.label
  };
});

var colors = _countries.default;
var expands = 0;

function renderItem(_ref) {
  var domProps = _ref.domProps,
      item = _ref.item,
      data = _ref.data,
      index = _ref.index;
  return _react.default.createElement("div", _objectSpread({}, domProps), item.id, " - ", domProps.children);
}

var renderScroller = function renderScroller(props) {
  delete props.tabIndex;
  return _react.default.createElement("div", _objectSpread({}, props));
};

var Demo = function (_Component) {
  (0, _inherits2.default)(Demo, _Component);

  var _super = _createSuper(Demo);

  function Demo(props) {
    var _this;

    (0, _classCallCheck2.default)(this, Demo);
    _this = _super.call(this, props);
    _this.state = {
      theme: 'default',
      maxTags: 3,
      text: '',
      index: 1,
      dataSource: colors,
      item: 'test',
      searchable: true,
      comboValue: 'RO',
      expanded: true,
      comboBoxValue: 'Algeria'
    };
    _this.loadLazyDataSource = _this.loadLazyDataSource.bind((0, _assertThisInitialized2.default)(_this));
    return _this;
  }

  (0, _createClass2.default)(Demo, [{
    key: "loadLazyDataSource",
    value: function loadLazyDataSource(_ref2) {
      var text = _ref2.text;
      console.warn('load@', text);
      var lowerText = text ? text = text.toLowerCase() : '';
      var filteredCountries = lowerText ? _countries.default.filter(function (country) {
        return country.label.toLowerCase().indexOf(lowerText) != -1;
      }) : _countries.default;
      return new Promise(function (resolve) {
        (0, _timers.setTimeout)(function () {
          return resolve(filteredCountries);
        }, 350);
      });
    }
  }, {
    key: "render",
    value: function render() {
      var _this2 = this;

      return _react.default.createElement("div", {
        className: "master_wrapper",
        style: {
          height: '50vh',
          paddingTop: 100
        }
      }, _react.default.createElement("div", {
        style: {
          marginBottom: 30
        }
      }, _react.default.createElement("button", {
        onClick: function onClick() {
          return _this2.setState({
            loading: !_this2.state.loading
          });
        }
      }, "toggle loading"), _react.default.createElement(_ComboBox.default, {
        style: {
          width: 150
        },
        multiple: true,
        itemEllipsis: true,
        dataSource: _countries.default,
        value: this.state.comboBoxValue,
        loading: this.state.loading,
        onChange: function onChange(comboBoxValue) {
          return _this2.setState({
            index: _this2.state.index + 1,
            comboBoxValue: comboBoxValue
          });
        }
      })), _react.default.createElement("div", {
        style: {
          transform: 'translate3d(0px, 0px, 0px)'
        }
      }, _react.default.createElement("div", {
        style: {
          border: '1px solid red',
          paddingTop: 500,
          paddingBottom: 200,
          width: 600,
          overflow: 'hidden'
        }
      }, _react.default.createElement(_ComboBox.default, {
        collapseOnSelect: true,
        constrainTo: function constrainTo(node) {
          return node.parentNode;
        },
        displayProperty: "label",
        style: {
          width: 400
        },
        positions: ['top', 'bottom'],
        dataSource: _countries.default
      }))), _react.default.createElement("input", {
        type: "checkbox",
        checked: this.state.searchable,
        onChange: function onChange(ev) {
          return _this2.setState({
            searchable: ev.target.checked
          });
        }
      }), "searchable", _react.default.createElement("br", null), _react.default.createElement(_ComboBox.default, {
        collapseOnSelect: true,
        changeValueOnNavigation: true,
        onChange: function onChange(value) {
          _this2.setState({
            theme: value
          });
        },
        placeholder: "AAA",
        defaultValue: null,
        dataSource: themes
      }), _react.default.createElement("br", null), _react.default.createElement("button", {
        onClick: function onClick() {
          _this2.setState({
            dataSource: function dataSource() {
              return new Promise(function (resolve) {
                (0, _timers.setTimeout)(function () {
                  return resolve(_countries.default);
                }, 3000);
              });
            }
          });
        }
      }, "load data source"), _react.default.createElement("br", null), _react.default.createElement("br", null), _react.default.createElement("div", {
        id: "constrain",
        style: {
          height: 300,
          border: '1px dashed gray',
          overflow: 'auto'
        }
      }, _react.default.createElement("div", {
        style: {
          height: 140
        }
      }), _react.default.createElement(_ComboBox.default, {
        rtl: false,
        constrainTo: "#constrain",
        style: {
          width: 'auto',
          position: 'relative'
        },
        renderItem: function renderItem(_ref3) {
          var domProps = _ref3.domProps;
          domProps.style = _objectSpread(_objectSpread({}, domProps.style), {}, {
            display: 'flex',
            flexFlow: 'row',
            alignItems: 'center'
          });
        },
        multiple: true,
        placeholder: "Enter Country",
        dataSource: this.state.dataSource
      })));
    }
  }]);
  return Demo;
}(_react.Component);

var _default = Demo;
exports.default = _default;
