// @inovua/reactdatagrid-community/BoolFilter/index.js
/** 
 * Copyright © INOVUA TRADING.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _createClass2 = _interopRequireDefault(require("@babel/runtime/helpers/createClass"));

var _classCallCheck2 = _interopRequireDefault(require("@babel/runtime/helpers/classCallCheck"));

var _assertThisInitialized2 = _interopRequireDefault(require("@babel/runtime/helpers/assertThisInitialized"));

var _inherits2 = _interopRequireDefault(require("@babel/runtime/helpers/inherits"));

var _possibleConstructorReturn2 = _interopRequireDefault(require("@babel/runtime/helpers/possibleConstructorReturn"));

var _getPrototypeOf2 = _interopRequireDefault(require("@babel/runtime/helpers/getPrototypeOf"));

var _defineProperty2 = _interopRequireDefault(require("@babel/runtime/helpers/defineProperty"));

var _react = _interopRequireDefault(require("react"));

var _CheckBox = _interopRequireDefault(require("../packages/CheckBox"));

var _debounce = _interopRequireDefault(require("../packages/debounce"));

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); enumerableOnly && (symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; })), keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = null != arguments[i] ? arguments[i] : {}; i % 2 ? ownKeys(Object(source), !0).forEach(function (key) { (0, _defineProperty2.default)(target, key, source[key]); }) : Object.getOwnPropertyDescriptors ? Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)) : ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } return target; }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = (0, _getPrototypeOf2.default)(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = (0, _getPrototypeOf2.default)(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return (0, _possibleConstructorReturn2.default)(this, result); }; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Boolean.prototype.valueOf.call(Reflect.construct(Boolean, [], function () {})); return true; } catch (e) { return false; } }

var BoolFilter = function (_React$Component) {
  (0, _inherits2.default)(BoolFilter, _React$Component);

  var _super = _createSuper(BoolFilter);

  function BoolFilter(props) {
    var _this;

    (0, _classCallCheck2.default)(this, BoolFilter);
    _this = _super.call(this, props);
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onChange", function (checked) {
      _this.onValueChange(checked);

      _this.setValue(checked);
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "componentDidUpdate", function () {
      if (_this.props.filterValue && _this.props.filterValue.value !== _this.state.value) {
        if (_this.props.filterValue) {
          var value = _this.props.filterValue.value;

          _this.setState({
            value: value
          });
        }
      }
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "setValue", function (checked) {
      _this.setState({
        value: checked
      });
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "onValueChange", function (checked) {
      _this.props.onChange && _this.props.onChange(_objectSpread(_objectSpread({}, _this.props.filterValue), {}, {
        value: checked
      }));
    });
    (0, _defineProperty2.default)((0, _assertThisInitialized2.default)(_this), "render", function () {
      var _this$props = _this.props,
          readOnly = _this$props.readOnly,
          filterEditorProps = _this$props.filterEditorProps;
      var finalEditorProps = typeof filterEditorProps === 'function' ? filterEditorProps(_this.props) : filterEditorProps;
      return _this.props.render && _this.props.render(_react.default.createElement(_CheckBox.default, _objectSpread(_objectSpread({}, finalEditorProps), {}, {
        readOnly: readOnly,
        theme: _this.props.theme,
        disabled: _this.props.disabled,
        onChange: _this.onChange,
        supportIndeterminate: true,
        indeterminateValue: null,
        className: "InovuaReactDataGrid__column-header__filter InovuaReactDataGrid__column-header__filter--bool",
        checked: _this.state.value
      })));
    });
    var filterValue = props.filterValue;
    _this.state = {
      value: filterValue ? filterValue.value : null
    };
    _this.onChange = _this.onChange.bind((0, _assertThisInitialized2.default)(_this));
    _this.onValueChange = _this.onValueChange.bind((0, _assertThisInitialized2.default)(_this));

    if (props.filterDelay && props.filterDelay >= 1) {
      _this.onValueChange = (0, _debounce.default)(_this.onValueChange, props.filterDelay, {
        leading: false,
        trailing: true
      });
    }

    return _this;
  }

  return (0, _createClass2.default)(BoolFilter);
}(_react.default.Component);

var _default = BoolFilter;
exports.default = _default;
